frontend/packages/operator-lifecycle-manager/src/components/create-operand.tsx

```diff
@@ -104,297 +178,465 @@ const defaultValueFor = (field: OperandField) => {
-                capabilities: [
-                  ...new Set(field.capabilities).add(
-                    SpecCapability.fieldGroup.concat(
-                      field.path.split('.')[0],
-                    ) as SpecCapability.fieldGroup,
-                  ),
-                ],
-              }
-            : field,
-        ),
-    [openAPI, providedAPI, clusterServiceVersion],
-  );
+  // Map providedAPI spec descriptors and openAPI spec properties to OperandField[] array
+  const initialFields: OperandField[] = React.useMemo(() => {
+    // Get fields from openAPI
+    const openAPIFields = fieldsForOpenAPI(openAPI, providedAPI, buffer);
+
+    // Get fields from providedAPI that do not exist in the OpenAPI spec.
+    const providedAPIFields = _.reduce(
+      providedAPI?.specDescriptors || [],
+      (providedAPIFieldsAccumulator, specDescriptor) => {
+        // If this field was already created, ignore it.
+        if (_.find(openAPIFields, { path: specDescriptor.path })) {
+          return providedAPIFieldsAccumulator;
+        }
+
+        // Add the field if it doesn't exist
+        return [...providedAPIFieldsAccumulator, ...specDescriptorToFields(specDescriptor, buffer)];
+      },
+      [],
+    );
 
-  const [fields, setFields] = React.useState<OperandField[]>(fieldsFromProps);
+    // Concatenate all fields and return
+    return [...openAPIFields, ...providedAPIFields];
+  }, [openAPI, providedAPI, buffer]);
+  const [fields, setFields] = React.useState<OperandField[]>(initialFields);
+
+  const initialFormValues = React.useMemo(() => {
+    const initialMetadataValues = {
+      'metadata.name': _.get(buffer, 'metadata.name', 'example'),
+      'metadata.labels': SelectorInput.arrayify(_.get(buffer, 'metadata.labels')) || [],
+    };
+    return fields.reduce<FormValues>((formValuesAccumulator, field) => {
+      const value = _.get(buffer, `spec.${field.path}`);
+      return {
+        ...formValuesAccumulator,
+        ...(value && { [field.path]: value }),
+      };
+    }, initialMetadataValues);
+  }, [fields, buffer]);
 
-  type FormValues = { [path: string]: any };
+  const [formValues, setFormValues] = React.useState<FormValues>(initialFormValues);
+  const [error, setError] = React.useState<string>();
+  const [formErrors, setFormErrors] = React.useState<FieldErrors>({});
 
-  const defaultFormValues = React.useMemo(
-    () =>
-      fields.reduce<FormValues>(
-        (allFields, field) => ({ ...allFields, [field.path]: defaultValueFor(field) }),
-        {},
+  // Group fields into advanced, arrayFieldGroup, fieldGroup, and normal fields for rendering.
+  // Note that arrayFieldGroup and fieldGroup fields are still flat after this. The memoized
+  // 'arrayFieldGroups' and 'fieldGroups' (below) further organizes these by their respective
+  // group descriptors to simplify rendering and state management.
+  const [advancedFields, arrayFields, groupFields, normalFields]: OperandField[][] = React.useMemo(
```

Comment:
Memoized grouping of fields so that we don't have to do a bunch of filtering logic when rendering.